[FEATURE]:  Extract latest fuel price from JORAM's PDF repository #10

HarryVasanth opened the pull request:
### 💁‍♂️ Please feel free to improve and push changes to this PR. 

### Related issues

Closes #10 

### What does this PR do?

> <sub>Clearly state the reason for this change. What problem is it solving or feature is it adding?</sub>

This PR completes @13dev's work on the fuel price extraction from JORAM's [PDF Repository](https://joram.madeira.gov.pt/joram/2serie/).

### Solution Overview

> <sub>Provide an overview of the solution implemented in this pull request. This should be a high level overview without getting into technical details. If applicable, include screenshots of UI or use GitHub compliant [`mermaid`](https://mermaid.live/) graphs to visually represent the solution.</sub>

- Get JORAM URL for current year's PDFs.
- Get the list of URLs ending with `.pdf`.
- Sort the list to get the latest PDF by date.
- Extract the fuel price from the PDF / Return empty string in case of no presence of fuel price.

### Implementation Details

> <sub>Explain the details of the implementation and the reasoning behind it. Where should the reviewer start?</sub>

Every line after `# FIXME (DEBUG):` can be commented out by whomever takes over the integration part. 
I vote for @Dntfreitas 😉

#### Outputs
Currently, it prints the following:
1. Latest PDF filename present in the repository :
`IISerie-018-2023-01-25.pdf`
2. Fuel price from the latest PDF:
`{}`
3. Fuel price from the [`Debug PDF`](https://joram.madeira.gov.pt/joram/2serie/Ano%20de%202022/IISerie-138-2022-07-22Supl.pdf):
`{'Gasolina super sem chumbo IO 95': '1,889', 'Gasóleo rodoviário': '1,789', 'Gasóleo colorido e marcado': '1,425'}`

joaoofreitas commented on joram.py:
```diff
@@ -1,24 +1,46 @@
+    ),
+)
+newest_pdf_link = sorted_pdf_links[-1]
+newest_pdf_filename = newest_pdf_link["href"].split("/")[-1]
```
We need to scrape them all to filter only the ones with matches. Then sort them. 

Then we can get it exactly like this.

The current output is this one:
We need to do this because not all files have the info we need to extract.
```sh
IISerie-019-2023-01-26.pdf
{}
```

HarryVasanth replied:
I was wondering if this runs daily, and it keeps looking for the prices on the latest PDF, then we wouldn't have to scrape all the PDFs, just the latest one. So if it has price, we update it. If it doesn't, we don't update.

carlosrsabreu replied:
The fuel prices are usually added to JORAM on Thursdays / Fridays. 

@HarryVasanth what happens if for some reason they add two PDFs at the same time? Only the last one will be scrapped?

joaoofreitas replied:
Hmm. That's a good point. 

But it's supposed to run every Friday, because JORAM updates a new file every Friday.

But yeah, it's a good solution after all 👀

HarryVasanth replied:
@carlosrsabreu: If it's on the same day, it will scrape the most recent one. So, one thing we could do is to increase the frequency of scraping to every 15 mins. So even if they make amendments to the price, we'll push the latest possible one. 

Currently, we are scraping at these two frequencies:

**Every 6 hours.**
```yml
cron: "0 */6 * * *"
 ```
 
**Every 30 mins on Fri, Sat, Sun, and Mon.**
```yml 
cron: "*/30 * * * 5,6,0,1"
```

joaoofreitas commented on joram.py:
```diff
@@ -1,24 +1,46 @@
+# FIXME (DEBUG) Print se the filename to scrape the content
+print(newest_pdf_filename)
+# FIXME (DEBUG): Working link to Joram PDF
+debug_pdf_joram = "https://joram.madeira.gov.pt/joram/2serie/Ano%20de%202022/IISerie-138-2022-07-22Supl.pdf"
```
_Notes to future implementations:_ 

- Remove Debug info and maybe create a python test file;
- Encapsulate this code inside a callable function.

@carlosrsabreu

carlosrsabreu replied:
Created the issue, thank you for the suggestion! 🙏

https://github.com/carlosrsabreu/devo-abastecer/issues/30

joaoofreitas reviewed (changes requested):
High quality code as expected @HarryVasanth!

But ___JORAM___ is unpredicable and there's some points missing here.

 > We should filter the list of files further because _not all files contain the extraction data_.

A non creative solution would be:
- [X] Getting the full list of _pdf_ links _(As you are doing right now)_;
- [X] Sort it by date ;
- [ ] Slice the list to have only 10-15 items only _(Should improve the scraping time i doubt they release more than that in a week)_;
- [ ] Then we scrape them all and remove the ones with a number of matches <= 0;
- [ ] From that, we can get the first _(or last)_ depending on sorting type _(ASC or DSC)_.

Right now I will travel for a few days, but if you want we can check this out together.

Do you guys have something to add?
@Dntfreitas @carlosrsabreu 

Excellent work @HarryVasanth quality code and great readability!
Let me know what do you think :)

carlosrsabreu reviewed (commented):
@HarryVasanth first of all, thank you a lot for the initiative in this PR!

I've tested it with some PDFs and it works perfectly!

A few notes/suggestions:
- We can run this script daily, but I would suggest running it with more frequency on Thursdays/Fridays, which are the days they usually add the fuel prices for the next week;
- I'm not sure if the idea was to trigger this file every time a new file is added to JORAM (was a perfect solution so we would retrieve the data as fast as possible), but for now, we can run the script and look for the last gas prices in files of the current week;
- After this (we can do this in other PR), we should then guarantee that we didn't scrap a PDF from the last week, which could public wrong gas prices.

HarryVasanth commented:
@carlosrsabreu @joaoofreitas
Updated the code and added some comments, so whomever will take over can skim through quickly. 
As we discussed, it iterates through the links.

Current output prints the following:
```zsh
# Prices
{'Gasolina super sem chumbo IO 95': '1,608', 'Gasóleo rodoviário': '1,488', 'Gasóleo colorido e marcado': '1,151'}

# Filename for Debug
IISerie-020-2023-01-27Supl.pdf

# URL for Debug
https://joram.madeira.gov.pt/joram/2serie/Ano%20de%202023/IISerie-020-2023-01-27Supl.pdf
```

joaoofreitas commented:
Perfect... I can't review it right now unfortunately.

@carlosrsabreu @Dntfreitas, please have the courtesy.

Great work @HarryVasanth, all your work is very valuable.

carlosrsabreu commented on joram.py:
```diff
@@ -44,23 +42,28 @@ def get_pdf_content_lines(pdf_raw_data):
+    while len(sorted_pdf_links) > 0:
+        newest_pdf_link = sorted_pdf_links.pop()
+        newest_pdf_filename = newest_pdf_link["href"].split("/")[-1]
+        newest_pdf_joram = joram_url + current_year + "/" + newest_pdf_filename
```
I think that we just need to guarantee that the newest pdf it's from the week we're looking for, to prevent finding and publishing the gas prices from the previous week. Am I thinking right, @HarryVasanth?

HarryVasanth replied:
Yes. For that, we only publish if the price has changed, plus we'd need to extract the dates. 
So if one of you guys will integrate this, then these things needs to be considered. 😅

carlosrsabreu reviewed (changes requested):
After making sure we're retrieving the gas prices for the following week, this PR is ready IMO!

Superb, @HarryVasanth!

HarryVasanth commented:
> After making sure we're retrieving the gas prices for the following week, this PR is ready IMO!
> 
> Superb, @HarryVasanth!

We can try to run both parallelly (writing to separate `.json` files). Once we iron out all the logical issues, bugs, etc. then we can go ahead to use this as our main form of extraction and keep the DRETT scraper as the backup.

carlosrsabreu commented:
> > After making sure we're retrieving the gas prices for the following week, this PR is ready IMO!
> > Superb, @HarryVasanth!
> 
> We can try to run both parallelly (writing to separate `.json` files). Once we iron out all the logical issues, bugs, etc. then we can go ahead to use this as our main form of extraction and keep the DRETT scraper as the backup.

Yes, I agree with that. So, we can merge this one and then create a PR to handle what you mentioned, right?

13dev reviewed (approved):
i would make a module for this (where we have a parser, tokenizer and maybe a scrapper) gj overall, would be nice to transform this in an api maybe?

carlosrsabreu commented:
I think we can merge this one. Then, we can try to optimize all the pieces and make this bot work even better.

Dntfreitas reviewed (approved):
Good Job! I just changed the use of constants. Also, there was some constant not being used.
